src/main/java/com/emc/object/s3/S3SignerV4.java

```diff
@@ -0,0 +1,392 @@
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.*;
+
+
+public class S3SignerV4 extends S3Signer {
+    private static final String HEADER_DATE_FORMAT = "EEE, dd MMM yyyy HH:mm:ss Z";
+    //The timestamp must be in UTC and in the following ISO 8601 format: YYYYMMDD'T'HHMMSS'Z'
+    private static final String AMZ_DATE_FORMAT = "yyyyMMdd'T'HHmmss'Z'";
+    private static final String AMZ_DATE_FORMAT_SHORT = "yyyyMMdd";
+    private static final long PRESIGN_URL_MAX_EXPIRATION_SECONDS = 60 * 60 * 24 * 7;
+    private static final SortedSet<String> excludedSignedHeaders = new TreeSet(Arrays.asList(
+        "authorization"
+    ));
+
+    public S3SignerV4(S3Config s3Config) {
+        super(s3Config);
+    }
+
+    public void sign(ClientRequest request, String resource, Map<String, String> parameters, Map<String, List<Object>> headers) {
+        // # Preparation, add x-amz-date and host headers
+        String date = null;
+
+        String serviceType = getServiceType();
+        StringBuilder hostHeader = new StringBuilder(request.getURI().getHost());
+
+        // If default port is used, do not include the port in host header
+        if(!(s3Config.getProtocol().equals("https") && request.getURI().getPort() == 443) &&
+                !(s3Config.getProtocol().equals("http") && request.getURI().getPort() == 80))
+            hostHeader.append(":").append(request.getURI().getPort());
+        if (headers.containsKey(S3Constants.AMZ_DATE)) {
+            date = RestUtil.getFirstAsString(headers, S3Constants.AMZ_DATE);
+        }
+        else {
+            date = getDate(parameters, headers);
+            request.getHeaders().add(S3Constants.AMZ_DATE, date);
+            request.getHeaders().add(RestUtil.HEADER_HOST, hostHeader);
+        }
+
+        String shortDate = getShortDate(date);
+
+        RestUtil.putSingle(headers, S3Constants.AMZ_DATE, date);
+        RestUtil.putSingle(headers, RestUtil.HEADER_HOST, hostHeader);
+        headers = request.getHeaders();
+
+        // #1 Create a canonical request for Signature Version 4
+        String canonicalRequest = getCanonicalRequest(request, parameters, headers);
+
+        // #2 Create a string to sign for Signature Version 4
+        // get HashedCanonicalRequest
+        byte[] hash = hash256(canonicalRequest);
+        String hashedRequest = hexEncode(hash);
+        SortedMap<String, String> canonicalizedHeaders = getCanonicalizedHeaders(headers, parameters);
+        StringBuilder signedHeaders = new StringBuilder();
+        for (String name : canonicalizedHeaders.keySet()) {
+            if (signedHeaders.length() != 0)
+                signedHeaders.append(";");
+            signedHeaders.append(name);
+        }
+        String stringToSign = getStringToSign(request.getMethod(), resource, parameters, headers, date, serviceType) + hashedRequest;
```

Comment:
A decent abstract would not require appending "hashedRequest" but a complete sign string returned by the function. Is it possible to leverage s3Config.isUseV2Signer() and put everything inside the method in parent class?

Reply:
hashedRequest is requied in v4 stringToSign. Have integrated hashedRequest into stringToSign for v4 signer.